lib/ecto/type.ex

```diff
@@ -1209,6 +1209,19 @@ defmodule Ecto.Type do
     false
   end
 
+  @doc """
+  Format type for error messaging and logs.
+  """
+  def format({:parameterized, type, params}) do
+    if function_exported?(type, :format, 1) do
+      apply(type, :format, [params])
+    else
+      "##{inspect(type)}<#{inspect(params)}>"
+    end
+  end
+
+  def format(type), do: inspect(type)
```

Comment:
We should handle composite types. For example, `{:array, {:parameterized, Ecto.Enum, params}` should be formatted as `{:array, ##Ecto.Enum<...>}`